File: 2015/day5/5.py
from typing import Dict
from common.utils import problem_harness, timeit, read_input


def has_double(s: str) -> bool:
  for i in range(1, len(s)):
    if s[i] == s[i - 1]:
      return True
  return False


def three_vowels(s: str) -> bool:
  return sum(map(s.count, list('aeiou'))) >= 3


def no_bad_strings(s: str) -> bool:
  return not any(bad in s for bad in ['ab', 'cd', 'pq', 'xy'])


def is_nice(s: str) -> int:
  return int(has_double(s) and three_vowels(s) and no_bad_strings(s))


def letter_sandwich(s: str) -> bool:
  """ xyx, or aaa, or efe """
  if len(s) < 3: # no sandwich possible
    return False

  for i in range(1, len(s) - 1):
    if s[i - 1] == s[i + 1]:
      return True
  return False


def two_pairs(s: str) -> bool:
  """ non overlapping pairs, xyxy or aaalskdjflskdjaa, not aaa"""
  for i in range(1, len(s)- 1):
    if s.count(s[i - 1:i + 1], i + 1):
      return True
  return False


def is_nice_part_2(s: str) -> int:
  return int(two_pairs(s) and letter_sandwich(s))


@timeit
def part1(filename: str) -> int:
  return sum(map(is_nice, read_input(filename)))


@timeit
def part2(filename: str) -> int:
  return sum(map(is_nice_part_2, read_input(filename)))


def main():
  problem_harness(part1, part2)

if __name__ == '__main__':
  main()

File: 2015/day5/5.go
package main


import (
	"strings"
	"aoc/utils"
)


func has_double_letter(s string) bool {
	for i := 1; i < len(s); i++ {
		if s[i] == s[i-1] {
			return true
		}
	}
	return false
}


func has_vowels(s string) bool {
	vowels := "aeiou"
	count := 0
	for _, c := range s {
		if strings.Contains(vowels, string(c)) {
			count++
		}
	}
	return count >= 3
}


func no_bad_strings(s string) bool {
	bad_strings := []string{"ab", "cd", "pq", "xy"}
	for _, bs := range bad_strings {
		if strings.Contains(s, bs) {
			return false
		}
	}
	return true
}


func has_letter_sandwich(s string) bool {
	// aaa, xyx, efe,...
	if len(s) < 3 {
		return false // not enough letters for a sandwich
	}
	for i := 1; i < len(s)-1; i++ {
		if s[i-1] == s[i+1] {
			return true
		}
	}
	return false
}


func has_double_pair(s string) bool {
	// if there are two non-overlapping pairs of letters
	if len(s) < 4 {
		return false
	}
	for i := 0; i < len(s)-2; i++ {
		if strings.Contains(s[i+2:], s[i:i+2]) {
			return true
		}
	}
	return false
}


func is_nice_part1(s string) bool {
	return has_double_letter(s) && has_vowels(s) && no_bad_strings(s)
}


func is_nice_part2(s string) bool {
	return has_double_pair(s) && has_letter_sandwich(s)
}


func part1(data string) int {
	return utils.Reduce(
		strings.Split(data, "\n"),
		func(accum int, cur string) int {
			if is_nice_part1(cur) {
				return accum + 1
			}
			return accum
		}, 0)
}


func part2(data string) int {
	return utils.Reduce(
		strings.Split(data, "\n"),
		func(accum int, cur string) int {
			if is_nice_part2(cur) {
				return accum + 1
			}
			return accum
		}, 0)
}


func main() {
	utils.ProblemSetUp(part1, part2)
}